tests/integration/test_charm.py

```diff
@@ -118,15 +128,60 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
+    """Test data is replicated to new units after a scale up."""
+    # Set a composite application name in order to test in more than one series at the same time.
+    application_name = build_application_name(series)
+
+    # Ensure the initial number of units in the application.
+    initial_scale = len(UNIT_IDS)
+    await scale_application(ops_test, application_name, initial_scale)
+
+    # Scale down the application.
+    await scale_application(ops_test, application_name, initial_scale - 1)
+
+    # Ensure the member was correctly removed from the cluster
+    # (by comparing the cluster members and the current units).
+    await check_cluster_members(ops_test, application_name)
+
+    # Scale up the application (2 more units than the current scale).
+    await scale_application(ops_test, application_name, initial_scale + 1)
+
+    # Assert the correct members are part of the cluster.
+    await check_cluster_members(ops_test, application_name)
 
 
-def convert_records_to_dict(records: List[tuple]) -> dict:
-    """Converts pyscopg2 records list to a dict."""
-    dict = {}
-    for record in records:
-        # Add record tuple data to dict.
-        dict[record[0]] = record[1]
-    return dict
+@pytest.mark.parametrize("series", SERIES)
+async def test_persist_data_through_leader_deletion(ops_test: OpsTest, series: str):
+    """Test data persists through a leader deletion."""
+    # Set a composite application name in order to test in more than one series at the same time.
+    application_name = build_application_name(series)
+    any_unit_name = ops_test.model.applications[application_name].units[0].name
+    primary = await get_primary(ops_test, any_unit_name)
+    password = await get_postgres_password(ops_test, primary)
+
+    # Write data to primary IP.
+    host = await get_unit_address(ops_test, primary)
+    logger.info(f"connecting to primary {primary} on {host}")
+    with db_connect(host=host, password=password) as connection:
+        connection.autocommit = True
+        connection.cursor().execute("CREATE TABLE leaderdeletiontest (testcol INT );")
+
+    # Remove one unit.
+    await ops_test.model.destroy_units(
+        primary,
+    )
+    await ops_test.model.wait_for_idle(apps=[application_name], status="active", timeout=1000)
+
+    # Add the unit again.
+    await ops_test.model.applications[application_name].add_unit(count=1)
+    await ops_test.model.wait_for_idle(apps=[application_name], status="active", timeout=1000)
+
+    # Testing write occurred to every postgres instance by reading from them
+    for unit in ops_test.model.applications[application_name].units:
+        host = unit.public_address
+        logger.info("connecting to the database host: %s", host)
+        with db_connect(host=host, password=password) as connection:
+            # Ensure we can read from "leaderdeletiontest" table
+            connection.cursor().execute("SELECT * FROM leaderdeletiontest;")
```

Comment:
add checks for when the primary gets removed that the configuration is accurate (ie the juju unit ips match the PG ips)

Reply:
Added on https://github.com/canonical/postgresql-operator/pull/14/commits/6fe523270b2b5a4424f9ed91c7038de88e66812c (and also on https://github.com/canonical/postgresql-operator/pull/14/commits/b23967277d37d7d2291c48fe0653f71954b09f91, that contains additional fixes).